continunet/network/unet.py

```diff
@@ -0,0 +1,135 @@
+        dropout_layer = Dropout(self.dropout)
+
+        return convolutional_block, dropout_layer(max_pooling_layer(convolutional_block))
+
+    def decoding_block(self, input_tensor, concat_tensor, filters, kernel_size=3):
+        """Decoding block for UNet."""
+        transpose_convolutional_layer = Conv2DTranspose(
+            filters, (3, 3), strides=(2, 2), padding="same"
+        )
+        skip_connection = Concatenate()(
+            [transpose_convolutional_layer(input_tensor), concat_tensor]
+        )
+        dropout_layer = Dropout(self.dropout)
+        return self.convolutional_block(dropout_layer(skip_connection), filters, kernel_size)
+
+    def build_model(self):
+        """Build the UNet model."""
+        input_image = Input(self.input_shape, name="img")
+
+        # Encoder Path
+        first_convolutional_tensor, first_feature_map = self.encoding_block(
+            input_image, self.filters * 1
+        )
+        second_convolutional_tensor, second_feature_map = self.encoding_block(
+            first_feature_map, self.filters * 2
+        )
+        third_convolutional_tensor, third_feature_map = self.encoding_block(
+            second_feature_map, self.filters * 4
+        )
+        fourth_convolutional_tensor, fourth_feature_map = self.encoding_block(
+            third_feature_map, self.filters * 8
+        )
+        latent_convolutional_tensor = self.convolutional_block(
+            fourth_feature_map, filters=self.filters * 16
+        )
+
+        # Decoder Path
+        sixth_convolutional_tensor = self.decoding_block(
+            latent_convolutional_tensor, fourth_convolutional_tensor, self.filters * 8
+        )
+        seventh_convolutional_tensor = self.decoding_block(
+            sixth_convolutional_tensor, third_convolutional_tensor, self.filters * 4
+        )
+        eighth_convolutional_tensor = self.decoding_block(
+            seventh_convolutional_tensor, second_convolutional_tensor, self.filters * 2
+        )
+        ninth_convolutional_tensor = self.decoding_block(
+            eighth_convolutional_tensor, first_convolutional_tensor, self.filters * 1
+        )
+
+        outputs = Conv2D(1, (1, 1), activation="sigmoid")(ninth_convolutional_tensor)
+        model = Model(inputs=[input_image], outputs=[outputs])
+        return model
+
+    def compile_model(self):
+        """Compile the UNet model."""
+        Input(self.input_shape, name="img")
+        model = self.build_model()
+        model.compile(
+            optimizer=Adam(), loss="binary_crossentropy", metrics=["accuracy", "iou_score"]
+        )
+        return model
+
+    def decode_image(self):
+        """Returns images decoded by a trained model."""
+        model = self.compile_model()
+        if self.trained_model is None or self.image is None:
+            raise ValueError("Trained model and image arguments are required to decode image.")
+        if isinstance(self.image, np.ndarray) is False:
```

Comment:
I feel like this check happens too late? Surely types should be checked before you start the encoding?

Reply:
If you put these in the init, you can't build the model without an image. I don't know if this is a problem tbh, but maybe it's nice to be able to build the model without an image when developing for training? Maybe it's unnecessary to worry about this. Thoughts?